tservice/src/seal.rs

```diff
@@ -71,6 +71,77 @@ impl<T: AsRef<[u8]>> Seal<Vec<u8>> for UnsealedData<T> {
     }
 }
 
+pub trait Unseal<T> {
+    /// The size in bytes needed to hold the unsealed MAC text
+    fn mac_text_len(&self) -> Result<usize>;
+
+    /// The length (in bytes) needed to hold the decrypted text
+    fn decrypted_text_len(&self) -> Result<usize>;
+
+    /// Unseal the data in `self`
+    fn unseal(&self) -> Result<UnsealedData<T>>;
+}
+
+impl<T: AsRef<[u8]>> Unseal<Vec<u8>> for SealedData<T> {
+    fn mac_text_len(&self) -> Result<usize> {
+        let result = unsafe {
+            mc_sgx_tservice_sys::sgx_get_add_mac_txt_len(
+                self.as_ref().as_ptr() as *const sgx_sealed_data_t
+            )
+        };
+        match result {
+            // Per the documentation, UINT32_MAX indicates an error
+            u32::MAX => Err(Error::Unexpected),
+            size => Ok(size as usize),
+        }
+    }
+
+    fn decrypted_text_len(&self) -> Result<usize> {
+        let result = unsafe {
+            mc_sgx_tservice_sys::sgx_get_encrypt_txt_len(
+                self.as_ref().as_ptr() as *const sgx_sealed_data_t
+            )
+        };
+        match result {
+            // Per the documentation, UINT32_MAX indicates an error
+            u32::MAX => Err(Error::Unexpected),
+            size => Ok(size as usize),
+        }
+    }
+
+    fn unseal(&self) -> Result<UnsealedData<Vec<u8>>> {
```

Comment:
This necessarily means we're going to need an `Unseal::unseal<T>(&self, &dest: UnsealedData<T>)`, but...

```suggestion
    #[cfg(feature = "alloc")]
    fn unseal_vec(&self) -> Result<UnsealedData<Vec<u8>>> {
```

Reply:
I thought alloc  as a feature was only needed for things outside of the enclave.

If it's true we don't need separate alloc logic:
- should `Unsealed` be converted to a `Vec<u8>` instead of a generic?
- should the `Unseal` go away and be methods on the `Unsealed` struct?